### ISSUE TITLE
[Feature Request] Dynamically update grammar rules during generation

### ISSUE THREAD

**Issue by AlienKevin** [ggerganov/llama.cpp#4149]
# Prerequisites

Please answer the following questions for yourself before submitting an issue.

- [x] I [searched using keywords relevant to my issue](https://docs.github.com/en/issues/tracking-your-work-with-issues/filtering-and-searching-issues-and-pull-requests) to make sure that I am creating a new issue that is not already open (or closed).
- [x] I reviewed the [Discussions](https://github.com/ggerganov/llama.cpp/discussions), and have a new bug or useful enhancement to share.

# Feature Description

Allow **changing the grammar rules dynamically** during the generation of each token. Potentially provide a callback function at the end of each token to ask for a new grammar for the next token.

# Motivation

We are trying to use code llama to provide code completion for [our language](https://hazel.org/) which is similar to Elm. Since the model hasn't seen this language during training, it tends to make simple syntax mistakes. The grammar feature comes in handy for us and contraining the generation to the grammar of our language helps alleviate this problem.

However, we would like more control over the generation process, particularly to enforce variable scoping and type consistency to prevent undefined variables and incorrect types. We have **a language server that can supply a list of acceptable language tokens at the cursor** that are in scope and type-correct. However, there's no way for us to supply this information to llama.cpp because the grammar rules are static and loaded at start up.

It would be great if **llama.cpp provides a callback so our language server can supply an updated grammar for the next token dynamically**. We think this feature can benefit other languages as well by giving more controls over the semantics of the generated output.

# Minimal Motivating Example

The ability to update the grammar rule for variables during generation can tremendously improve generation quality. Our language server can supply a set of names that are in scope to prevent the language model from hallucinating undefined functions and variable names.

**Comment by WeirdConstructor** [comment]
Oh interesting. I was just thinking about swapping out the grammar between generating a sentence the other day. My use case would have been to restrict the possible tokens in an natural language questionnaire setting though.

**Comment by KerfuffleV2** [comment]
I'm pretty sure you can already do this via the API. Sampling already happens after each decode operation during generation. It's possible to use different grammars, or also directly control sampling yourself. It's not completely clear what you mean when you say "llama.cpp" - there's the API, there's various examples of using that are based on the API. There's also stuff like language bindings for Python (just as an example) which are based on the API.

**Comment by AlienKevin** [comment]
> It's not completely clear what you mean when you say "llama.cpp"

I'm primarily using the console for testing. Currently, my command looks roughly like this:
```sh
./main \
    --grammar-file grammar.gbnf \
    -f prompt.txt \
    -m codellama-34b.gguf
```

I think a convenient interface for console users would be an extra command-line argument like `grammar-callback` that takes a shell/python script instead of a static grammar file:
```sh
./main \
    --grammar-callback callback.sh \
    -f prompt.txt \
    -m codellama-34b.gguf
```
The `llama.cpp` file can then call into `callback.sh` during generation and construct a new grammar on the fly for the next token. Frequency or trigger condition can be adjusted but they would be extensions to this feature request. The callback should be given the current generation status and return a new grammar string for the next token. The current status should include at least the prompt and the generated output. An example callback script may look like:
```sh
#!/bin/bash
# Calls the language server with the prompt and the generated output
# The server returns a new grammar string to llama.cpp
./my_lang_server "$1" "$2"
```

**Comment by KerfuffleV2** [comment]
My personal opinion is that if you're at the point where you want to callback every token that you're better off using llama.cpp as an API rather than running the examples from the console. It wouldn't be impossible to add a feature like that to `main` but it's a very, very specific, narrow use case.

If you're not comfortable with C/C++ stuff there's various bindings to llama.cpp including Python. If you already have a list of valid tokens each time you're doing generation, you probably also don't even need the complexity of the grammar stuff. Just ban all the tokens in the model that don't conform to your grammar and let normal sampling choose from what remains.

**Comment by WeirdConstructor** [comment]
> I'm pretty sure you can already do this via the API.

At first I thought the issue was about the API and not the `main` binary or spinning up extra processes to reconfigure the sampler.

My comment was about the `common/sampling.h` API. There is no function call to replace the grammar in the sampling context from the C/C++ API. Except maybe monkey patching the grammar into the sampling context somehow - and pray it doesn't break in future. Of course, there is always the possibility of using `llama_sample_grammar()` yourself and re-implementing the specific sampling algorithm you need.

**Comment by KerfuffleV2** [comment]
@WeirdConstructor 

> There is no function call to replace the grammar in the sampling context from the C/C++ API. 

In `common/sampling.h`:

```c++
// Copy the sampler context
void llama_sampling_cp(llama_sampling_context * src, llama_sampling_context * dst)
```

You can just build a new sampling context with the grammar you want (or build a set of them at the start if you already know what grammars you'll need in advance). Then you can `cp` the context you want to use onto the current one and sample normally.

There's also a function to reset the grammar. I'm not completely sure how OP's stuff works so they might need to do extra stuff to try to keep the grammars in sync with the current state. It sounded like that may not be necessary if they know _exactly_ what tokens they want to accept at a certain point.

I'm still not sure the grammar sampling is really the right tool for the job though, since you basically use it for more general rules. If they already know exactly which tokens are valid at the point they're sampling then there's a faster/simpler approach which I already mentioned.

**Comment by AlienKevin** [comment]
> If they already know exactly which tokens are valid at the point they're sampling then there's a faster/simpler approach which I already mentioned.

I think we know the exact set of tokens at a limited number of positions but we still need simple regex/cfg rules for most positions.

> They can just copy a function like llama_sampling_sample and make whatever modifications are necessary for their specific use case.
`llama_sampling_sample` looks promising for my application, though I'm not too sure how to update the grammar:

1. In the `llama_sampling_sample` function, there's [a place](https://github.com/ggerganov/llama.cpp/blob/0b871f1a04ef60e114bbe43004fd9c21114e802d/common/sampling.cpp#L168-L170) where `llama_sample_grammar` is called. Suppose I know the grammar rule for the next token, do I just initialize a new grammar object and pass it to `llama_sample_grammar` instead of using `ctx_sampling->grammar`? One potential issue I see is that the grammar object seems to be keeping some internal states like the stacks. Would it be okay for me to reset the whole grammar in this sampling function?
2. How can I get the prompt + current output in the `llama_sampling_sample` function? I plan to call my language server in this function to retrieve a grammar.
3. I noticed that `llama_sampling_sample` is called in [`examples/main/main.cpp`](https://github.com/ggerganov/llama.cpp/blob/0b871f1a04ef60e114bbe43004fd9c21114e802d/examples/main/main.cpp#L626). Is the sampling function called on a token-by-token basis?

**Comment by KerfuffleV2** [comment]
> Suppose I know the grammar rule for the next token, do I just initialize a new grammar object 

I think what you'd want to do is build a set of the grammar states you expect to encounter at the beginning (before doing any inference), so you'd make a sampling context for each of those, plus one more for your current sampling state. Then when you're at the point where you want to use a specific grammar rule, you'd just copy that over your current sampling context.

If it was me, I don't think I would be using the grammar stuff for this though. I'd just grab the text for the model's vocabulary and ban any tokens that don't conform to what I want to sample at that point. That's kind of how the grammar stuff works internally, except it has a bunch more complicated stuff going on to be flexible enough to deal with an actual grammar.

> How can I get the prompt + current output in the llama_sampling_sample function?

The sampling context has a `prev` field which is a vector of the last tokens. Basically, everything that got accepted with `llama_sampling_accept`. I believe `llama_sampling_sample` automatically handles that when it samples a token - accepting it with the grammar if there is one and adding it to `prev`.

Note though that if you're copying contexts onto the current one and you want that to be populated you have two options: Keep _all_ your contexts up to date - so you'd call accept on all of them when you sample a token. Alternatively I guess you could walk through _all_ your current tokens and accept them after you copy one of the empty contexts onto the current one to bring it up to the current state.

> Is the sampling function called on a token-by-token basis?

Yes. The way sampling works is you evaluate the model and you get a set of logits. Then you sample the logits and end up with a token. Then you pass the token that you sampled to the model and evaluate the model again (evaluate is called `llama_decode` in llama.cppland).

In the case of the initial prompt, you feed the prompt in as a batch and and request logits for the _last_ token in that batch. Then you'll sample from that to do the first actual LLM-generated token.

**Comment by AlienKevin** [comment]
@KerfuffleV2 Thanks for the detailed response! For now, I will disregard efficiency and go with the simplest approach to see if it works.
